src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp

```diff
@@ -503,24 +504,161 @@ real Hydro_CheckMinEintInEngy( const real Dens, const real MomX, const real MomY
 
 
 //-------------------------------------------------------------------------------------------------------
-// Function    :  Hydro_CheckNegative
-// Description :  Check whether the input value is <= 0.0 (also check whether it's Inf or NAN)
+// Function    :  Hydro_CheckUnphysical
+// Description :  Case1  :  Check if the input value is <= 0.0           (also check whether it's Inf or NAN)
+//                          --> Cons and Prim must be NULL
+//                Case2  :  Check if conserved variables are unphysical  (also check the existence of root in SRHD)
+//                          --> Input and Prim must be NULL
+//                Case3  :  Check if primitive variables are unphysical
+//                          --> Input and Cons must be NULL
 //
-// Note        :  Can be used to check whether the values of density and pressure are unphysical
+// Note        :  Two of Cons, Prim, and Input must be NULL
+//                --> Can be used to check whether the values of density and pressure are unphysical
 //
-// Parameter   :  Input : Input value
+// Parameter   :  Cons      :  Conserved variables
+//                Prim      :  Primitive variables
+//                Info      :  Additional information
+//                Input     :  Input value
+//                Passive   :  Passive scalars
+//                File      :  __FILE__
+//                Function  :  __FUNCTION__
+//                Line      :  __LINE__
+//                Show      :  true  --> Show error message
+//                             false --> Show nothing
 //
-// Return      :  true  --> Input <= 0.0  ||  >= __FLT_MAX__  ||  != itself (Nan)
-//                false --> otherwise
+// Return      :  Case1  :  true  --> Input <= 0.0  ||  >= __FLT_MAX__  ||  != itself (Nan)
+//                          false --> otherwise
+//                Case2  :  true  --> conserved variables are unphysical or the root(SRHD only) do not exist 
+//                          false --> otherwise
+//                Case3  :  true  --> primitive variables are unphysical
+//                          false --> otherwise
 //-------------------------------------------------------------------------------------------------------
 GPU_DEVICE
-bool Hydro_CheckNegative( const real Input )
+bool Hydro_CheckUnphysical( const real Cons[], const real Prim[], const real* const Input, const real Passive[],
+                            const char Info[], const char File[], const char Function[], const int Line, bool Show )
```

Comment:
Thank you for suggesting using `switch()`! But I face the catch-22 of using `goto` and `switch` in the following scenario:

If `goto` is banned completely in GAMER, the criteria in the `case` statement for checking conserved variables will be something like below:
```
bool FailCell = false;

// check conserved variables
case 2:
         // check NaN
         if (  Fields[0] != Fields[0]
            || Fields[1] != Fields[1]
            || Fields[2] != Fields[2]
            || Fields[3] != Fields[3]
            || Fields[4] != Fields[4]  )            FailCell = true;
 
         // check +inf and -inf for velocities and negative for mass density and pressure
         if (  (real)  TINY_NUMBER >= Fields[0] || Fields[0]  >= (real)HUGE_NUMBER
            || (real) -HUGE_NUMBER >= Fields[1] || Fields[1]  >= (real)HUGE_NUMBER
            || (real) -HUGE_NUMBER >= Fields[2] || Fields[2]  >= (real)HUGE_NUMBER
            || (real) -HUGE_NUMBER >= Fields[3] || Fields[3]  >= (real)HUGE_NUMBER
            || (real)  TINY_NUMBER >= Fields[4] || Fields[4]  >= (real)HUGE_NUMBER )            FailCell = true;
 
         if ( FailCell && Verbose )
         {   
            printf( "ERROR: unphysical primitive variables at file <%s>, line <%d>, function <%s>\n",
                            File, Line, Function );
            printf( "       rho=%14.7e, Ux=%14.7e, Uy=%14.7e, Uz=%14.7e, P=%14.7e\n",
                            Fields[0], Fields[1], Fields[2], Fields[3], Fields[4] );
         }  

break;
```
In the context above, when `Field[] = nan`, `Filed[]` will go into the next `if` statement, leading to meaningless calculation. The more criteria involved, the more serious the issue. Thus, I prefer using `goto` over `if` inside the case statement for multiple criteria. How do you think?

Reply:
Or using `if ... else if` statement to handle _multiple_ critiria inside a _single_ `case`?

Reply:
@zengbs Using `if … else if` sounds like a reasonable solution.

Reply:
Okay, thanks for your suggestion.

Reply:
Regarding this quote:
> Add sanity checks of Cons, Prim, and SingleField when enabling GAMER_DEBUG. For example, validate Cons != NULL when Mode == UNPHY_MODE_CONS. This should also be put inside switch.

Once the single field, conserved and primitive variables are integrated in `Fields[]`, IIUC, we have no idea how many elements of `Fields[]` within `Hydro_CheckUnphysical()` is.

Either of the following syntaxes do not work within a function: 
Length of `Fields[]` = `sizeof(Fields)/sizeof(*Fields)`
Length of `Fields[]` = `*(&Fields + 1) - Fields`

If the sanity check is necessary here, passing the length of `Field[]` into `Hydro_CheckUnphysical()` may be the only possible way. What do you think?

@hyschive

Reply:
Ah! We can identify the length of `Fields[]` by `Mode`.

Reply:
IIUC, `NCOMP_FLUID` is always equal to the number of conserved and primitive variables even when dual energy, cosmic-ray, passive scalars are involved. If yes, we could assume the length of `Fields[]` = `NCOMP_FLUID` when `Mode` = `UNPHY_MODE_CONS/UNPHY_MODE_PRIV`. What do you think?

Reply:
- [x] Rename Input as SingleField (but also check the suggestion about adding Mode, which may eliminate Input at all).
- [x] Rename Info as SingleFieldName
- [x] Rename Show as Verbose

  - [x] Add another parameter named, for example, Mode, to explicitly specify the target case among the three. It will avoid ambiguity when, for example, both Cons and Prim are not NULL. Some related suggestions:

   - [x] Declare different Mode in Typedef.h
   - [x] Use switch instead of if ... else for different modes
   - [x] Put respective error messages of each mode inside switch instead of using goto GOTCHA
   - [x] Add sanity checks of Cons, Prim, and SingleField when enabling GAMER_DEBUG. For example, validate Cons != NULL when Mode == UNPHY_MODE_CONS. This should also be put inside switch.
   - [x] Integrate Cons, Prim, and Input.